internal/shared/README.md

```diff
@@ -0,0 +1,6 @@
+# Shared
```

Comment:
I plan to describe this approach here: https://github.com/open-telemetry/opentelemetry-go/blob/main/CONTRIBUTING.md#style-guide as part of https://github.com/open-telemetry/opentelemetry-go/issues/3846